onnxruntime/core/session/compile_api.cc

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#if !defined(ORT_MINIMAL_BUILD)
+
+#include "core/session/compile_api.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "core/framework/error_code_helper.h"
+#include "core/session/abi_session_options_impl.h"
+#include "core/session/inference_session.h"
+#include "core/session/model_compilation_options.h"
+#include "core/session/onnxruntime_session_options_config_keys.h"
+#include "core/session/ort_apis.h"
+#include "core/session/ort_env.h"
+#include "core/session/utils.h"
+
+#define ORT_C_API_RETURN_IF_ERROR(expr)                 \
+  do {                                                  \
+    auto _status = (expr);                              \
+    if ((!_status.IsOK())) return ToOrtStatus(_status); \
+  } while (0)
+
+using namespace onnxruntime;
+
+ORT_API(void, OrtCompileAPI::ReleaseModelCompilationOptions,
+        _Frees_ptr_opt_ OrtModelCompilationOptions* ort_model_compile_options) {
+  delete reinterpret_cast<onnxruntime::ModelCompilationOptions*>(ort_model_compile_options);
+}
+
+ORT_API_STATUS_IMPL(OrtCompileAPI::CreateModelCompilationOptionsFromSessionOptions, _In_ const OrtEnv* env,
+                    _In_ const OrtSessionOptions* session_options, _Outptr_ OrtModelCompilationOptions** out) {
+  API_IMPL_BEGIN
+  if (env == nullptr) {
+    return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "The env argument must be a non-null pointer");
+  }
+
+  if (session_options == nullptr) {
+    return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "The session_options argument must be a non-null pointer");
+  }
+
+  auto model_compile_options = std::make_unique<onnxruntime::ModelCompilationOptions>();
+  model_compile_options->session_options = std::make_unique<OrtSessionOptions>(*session_options);
+
+  model_compile_options->session_options->value.has_explicit_ep_context_gen_options = true;
+  model_compile_options->session_options->value.ep_context_gen_options = session_options->value.GetEpContextGenerationOptions();
+  model_compile_options->session_options->value.ep_context_gen_options.enable = true;
+  model_compile_options->session_options->value.ep_context_gen_options.overwrite_existing_output_file = true;
+  model_compile_options->session_options->value.ep_context_gen_options.error_if_no_compiled_nodes = true;
```

Comment:
Is there a way for the user to change these values? Or is that not required?

Reply:
We can add new API functions on OrtModelCompilationOptions to allow the user to set this in the future if needed. I wanted to limit the size of this PR.